cmd, eth, rpc: fix some RPC issues with pending blocks

karalabe opened the pull request:
This PR mainly fixes an important RPC call, namely getting the account transaction count (`GetTransactionCount`) in the pending block. AFAIK, this is used by web3.js to figure out the account nonce to send the subsequent transaction with.

A few other fixes include:
 * Logging error messages in the RPC tester.
 * Removing a few 0 fields from GetBlocks when the pending block is queried.
 * A minor fix in the API interface{} to hex converter which didn't handle nil's fully properly.

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Jan 27 16:06:06 UTC 2016_

obscuren commented:
:-1: This fixes only one specific case, namely the one mentioned in the linked issue. There are several other instances where this will fail:

* `eth_call`;
* `eth_getBalance`;
* `eth_storageAt`;
* `eth_getData`;
* `eth_getCode`;
* `debug_call` (once added);
* and at some point logs (when we add them).

All of the above functions can be queried using the `pending` state. This PR only fixes the one instance of `eth_getTransactionCount` (`nonce`). 

I propose that you add to this PR a helper function that will a state based on the block number. I also think it would be wise to do this in two stages; 1. the hotfix 2. refactor the code.

**PR-1**: Hotfixed state retrieval:

```go
func (api *PublicBlockChainAPI) getState(root common.Hash, blockNum rpc.BlockNumber) (*state.StateDb, error) {
    if blockNum == -2 /* or whatever the equivalent is */ {
        return api.miner.PendingState(), nil
    }

    return state.New(root, api.chainDb)
}
```

and basically search and replace **almost** all `state.New(...)` with this call (please note that not all need to change).

***

**PR-2**: Refactor state retrieval and error checking: change all of the `if block := blockByNumber(...); block != nil` is changed to something like:

 ```go
state, err := stateByNumber(...)
if err != nil {
    "", err
}
/* rest of the code */
```

Why I suggest this rather radical change is because all of the `blockByNumber` if statements is for the exact same purpose as listed above 1. check if the block actual exist 2. check we have the state, and we can do this in one go.

karalabe commented:
@obscuren PTAL

 * I've added a `stateAndBlockByNumber` method (the block itself is used too in the `doCall`). 
 * There was a eth_getData method, that basically just returned the contract code, and the only invocation was by eth_getCode. Given that eth_getData is not in the API specs even, I consider this a refactor artifact or some legacy code and removed it.

Three concerns:
 * The `doCall` had some weird call parametrization (the header used in the call was always the head header, not the one the call was invoked on), and there are a few hardcoded numbers in there. Could you please check that they make sense?
 * The official rpc-tests repo doesn't catch any of these issues that we're correcting here, so I'm guessing the tests there are probably stale and not used any more, which is a problem because we don't have a reference to validate against any more.
 * One test in there still manages to fail, though I can't explain why, nor do I know if the test is actually up to date or faulty itself.

```
  11) eth_getTransactionByBlockNumberAndIndex IPC should return transactions for the pending block when using "pending" and sending transactions before:
     Uncaught AssertionError: is object: expected null to be an object
      at Function.assert.isObject (node_modules/chai/lib/chai/interface/assert.js:528:35)
      at Object.499 (test/eth_getTransactionBy.js:24:20)
      at lib/ipcProvider.js:63:44
      at Array.forEach (native)
      at Socket.<anonymous> (lib/ipcProvider.js:47:47)
      at Socket.<anonymous> (_stream_readable.js:746:14)
      at emitReadable_ (_stream_readable.js:408:10)
      at emitReadable (_stream_readable.js:404:5)
      at readableAddChunk (_stream_readable.js:165:9)
      at Socket.Readable.push (_stream_readable.js:127:10)
      at Pipe.onread (net.js:526:21)
```

karalabe commented:
I've found another `pending` block issue. When a filter was constructed using a pending block, it resulted in a semi infinite database lookup (we iterated blocks 0 to max_uint64-2). In the old code this was handled in the decoder where the filter block numbers aren't allowed to be < 0 (apart form -1 for the latest). I've updated the filter api code to reflect this old behavior.

obscuren commented on eth/api.go:
```diff
@@ -321,16 +342,12 @@ func (s *PublicBlockChainAPI) BlockNumber() *big.Int {
-	state, err := state.New(block.Root(), s.chainDb)
-	if err != nil {
+	state, _, err := stateAndBlockByNumber(s.miner, s.bc, blockNr, s.chainDb)
+	if state == nil || err != nil {
```
Should it ever be the case that err is nil and state is nil?

obscuren replied:
Additionally if the state is nil and the err is nil it will still return a nil, nil

karalabe replied:
If the block's not found then state is nil. I'm not sure what the API expects in those cases, I've retained the existing behavior from @bas-vk 's code.

obscuren replied:
Okay. Perhaps we should return an error if a block is not found that we are trying to query results for.

obscuren commented on eth/api.go:
```diff
@@ -434,38 +450,26 @@ func (s *PublicBlockChainAPI) NewBlocks(args NewBlocksArgs) (rpc.Subscription, e
-		}
-		return common.ToHex(res), nil
+	state, _, err := stateAndBlockByNumber(s.miner, s.bc, blockNr, s.chainDb)
+	if state == nil || err != nil {
```
Same as above

obscuren commented on eth/filters/api.go:
```diff
@@ -239,13 +239,13 @@ func (args *NewFilterArgs) UnmarshalJSON(data []byte) error {
 	}
 
-	if raw.From == nil {
+	if raw.From == nil || raw.From.Int64() < 0 {
```
What does this fix exactly?

karalabe replied:
The API supports passing in valid block numbers + latest + pending. If you pass in latest (-1), that is handled by the filter correctly. If you pass in pending (-2) that's not handled in the filter, which will convert it to uint64 (huge number) and will start iterating the blockchain util that huge number.

karalabe replied:
This is specifically true for the ToBlock field. I'm not sure whether From needs it or not, but we definitely need to convert the API "special block numbers" to "filter numbers" properly.

obscuren commented:
:+1:
